### ISSUE TITLE
.inv() not correctly inverting `Operation.decomposition()`

### ISSUE THREAD

**Issue by cvjjm** [XanaduAI/pennylane#533]
Update: I thought this would only be a problem in the drawer but also the circuit does not yield the expected result.

A user (like myself) might want to define their own `Operation` and start from one of the PennyLane Operations, e.g., `Rot`.

Now, If I just copy the definition of `Rot` to my module and rename it to `MyRot`, then the operation is no longer natively supported by, e.g., the `default.qubit` device, but it has a decomposition (the same as that of `Rot` into supported gates) and so I can use it without problems in my qnode circuits. However problems start to occur when I use `MyRot(...).inv()`.

A circuit such as
```
@qml.qnode(dev)
def circuit(*params):
    MyRot(0.1, 0.2, 0.3, wires=0).inv()
    Rot(0.1, 0.2, 0.3, wires=0)
    return qml.probs(range(wires))
```
does not yield the expected result (the two gates should cancel) and `circuit.draw()` looks the same irrespective of whether the .inv() is there or not. If I put the `.inv()` on `Rot` instead of the `MyRot` then the correct result is obtained.

As expected, the `Rot` gate is displayed as such by the circuit drawer and `MyRot` is decomposed into elementary rotations, but the individual rotations do not get an `.inv` when MyRot is inverted and their order is not reversed.

**Comment by antalszava** [comment]
Hi @cvjjm, thanks so much for opening an issue for this! :blush: 

Could you please provide more details on how exactly the definition of `MyRot` looks like in your case? In specific, was it decorated with `@template`?

**Comment by josh146** [comment]
It sounds like `MyRot` is not a template, but a custom operation, can you confirm @cvjjm? It might be useful if you could post your minimal (non)-working example 🙂 

It sounds like this could be a bug in how either PennyLane or the `default.qubit` simulator is treating the inverses of *decomposed* operations!

**Comment by cvjjm** [comment]
I agree that for what I originally wanted to achieve I could just as well define a template, but I then found it more convenient to subclass Operation (for example because I can define num_params and num_wires) and in this way discovered what I think is indeed a bug in how PL (or default.qubit?) treats the inverse of decomposable operations. Inv on operations that need to be decomposed doesn't seem to do anything. Instead of queuing the inverse of the gates from the decomposition in the reverse order, the gates from the decomposition are applied normally.

`MyRot` is just of demonstration purposes and a plain vanilla copy of `Rot()` and consequently looks like this:
```
class MyRot(Operation):
    r"""MyRot(phi, theta, omega, wires)
    A simple copy or Rot() to demonstrate a problem.
    .. math::
        R(\phi,\theta,\omega) = RZ(\omega)RY(\theta)RZ(\phi)= \begin{bmatrix}
        e^{-i(\phi+\omega)/2}\cos(\theta/2) & -e^{i(\phi-\omega)/2}\sin(\theta/2) \\
        e^{-i(\phi-\omega)/2}\sin(\theta/2) & e^{i(\phi+\omega)/2}\cos(\theta/2)
        \end{bmatrix}.
    **Details:**
    * Number of wires: 1
    * Number of parameters: 3
    * Gradient recipe: :math:`\frac{d}{d\phi}f(R(\phi, \theta, \omega)) = \frac{1}{2}\left[f(R(\phi+\pi/2, \theta, \omega)) - f(R(\phi-\pi/2, \theta, \omega))\right]`
      where :math:`f` is an expectation value depending on :math:`R(\phi, \theta, \omega)`.
      This gradient recipe applies for each angle argument :math:`\{\phi, \theta, \omega\}`.
    .. note::
        If the ``Rot`` gate is not supported on the targeted device, PennyLane
        will attempt to decompose the gate into :class:`~.RZ` and :class:`~.RY` gates.
    Args:
        phi (float): rotation angle :math:`\phi`
        theta (float): rotation angle :math:`\theta`
        omega (float): rotation angle :math:`\omega`
        wires (Sequence[int] or int): the wire the operation acts on
    """
    num_params = 3
    num_wires = 1
    par_domain = "R"
    grad_method = "A"

    @staticmethod
    def _matrix(*params):
        a, b, c = params
        return RZ._matrix(c) @ (RY._matrix(b) @ RZ._matrix(a))

    @staticmethod
    def decomposition(phi, theta, omega, wires):
        decomp_ops = [RZ(phi, wires=wires), RY(theta, wires=wires), RZ(omega, wires=wires)]
        return decomp_ops
```
This together with the qnode above (and the usual boilerplate to set up a one wire device and such) is a minimal non, working example. Observe that the output of the above qnode is `[1,0]` in case the `.inv()` is on `Rot()` and it is not in case `.inv` is on `MyRot()`. Printing `circuit.draw()` reveals that `MyRot()` is deomposed but the resulting sequence of gates is not inverted.

**Comment by josh146** [comment]
Indeed looks like a bug in `Operation.decomposition()`! Thanks for alerting this to us @cvjjm

**Comment by johannesjmeyer** [comment]
Hi @cvjjm, this was fixed in PR #543. You can use the new functionality when directly installing the latest master branch of PennyLane.

**Comment by cvjjm** [comment]
Thank you very much for fixing this so fast!
